### PR TITLE
Show results count while building filter

### PR DESCRIPTION
**Pull request by saskliutas**
Closes https://github.com/iTwin/presentation/issues/221

Prerequisite https://github.com/iTwin/appui/pull/467

### CODE REVIEW

File: packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx [hunk 1]
```diff
@@ -168,6 +181,38 @@ function PresentationInstanceFilterDialogContent(props: PresentationInstanceFilt
...
+  const { value, inProgress } = useDebouncedAsyncValue(
+    useCallback(async () => {
+      const filter = buildFilter({ ignoreErrors: true });
+      if (!filter) {
+        return undefined;
+      }
+
+      try {
+        return await getResultsCount(filter);
+      } catch {}
```
Comment (grigasp) [inline packages/components/src/presentation-components/instance-filter-builder/PresentationInstanceFilterDialog.tsx:199]: Need to handle `PresentationError` with `ResultSetTooLarge` status.

File: packages/components/src/presentation-components/tree/controlled/PresentationTreeRenderer.tsx [hunk 2]
```diff
@@ -79,13 +84,28 @@ export function PresentationTreeRenderer(props: PresentationTreeRendererProps) {
...
+
+  const getFilteredResultsCount = useCallback(
+    async (filter: PresentationInstanceFilterInfo) => {
+      const instanceFilter = await convertToInstanceFilterDefinition(filter.filter, imodel);
+      return Presentation.presentation.getNodesCount({
+        imodel,
+        rulesetOrId: dataProvider.rulesetId,
+        instanceFilter,
+        parentKey: filterNode.key,
+      });
```
Comment (grigasp) [inline packages/components/src/presentation-components/tree/controlled/PresentationTreeRenderer.tsx:105]: need to pass `hierarchyLevelSizeLimit`

File: packages/components/src/presentation-components/tree/controlled/PresentationTreeRenderer.tsx [hunk 3]
```diff
@@ -23,6 +26,7 @@ import { useHierarchyLevelFiltering } from "./UseHierarchyLevelFiltering";
 export interface PresentationTreeRendererProps extends TreeRendererProps {
   imodel: IModelConnection;
   modelSource: TreeModelSource;
+  dataProvider: IPresentationTreeDataProvider;
```
Comment (grigasp) [inline packages/components/src/presentation-components/tree/controlled/PresentationTreeRenderer.tsx:29]: I'm really not a fan of passing a data provider to the renderer. Let's pass a node loader here, as discussed offline.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: packages/components/src/presentation-components/tree/PresentationTreeNodeItem.ts [hunk 4]
```diff
@@ -19,9 +25,9 @@ export interface PresentationTreeNodeItemFilteringInfo {
    * Descriptor that describes instances of this tree node item hierarchy level. It can be used to create instance
    * filter using [[PresentationInstanceFilterBuilder]].
    *
-   * It can be set to a function in order to lazy load it.
+   * If it is set to `undefined` hierarchy level under this node is not filterable.
    */
-  descriptor: Descriptor | (() => Promise<Descriptor>);
+  descriptor?: HierarchyLevelFilteringDescriptor;
```
[the remark this answers is not in the document]
Comment (saskliutas) [reply packages/components/src/presentation-components/tree/PresentationTreeNodeItem.ts]: Removed, this was a left over from initial implementation

**Review by grigasp — APPROVED** [review]

**Pull request merged**